### PR TITLE
Integrate app updater with UI

### PR DESCRIPTION
**Pull request by gzdunek**
Part 5 of https://github.com/gravitational/teleport/issues/50948. [RFD](https://github.com/gravitational/teleport/blob/master/rfd/0144-client-tools-updates.md#teleport-connect-automatic-updates-added-on-2025-07-10-by-gzdunek).

This PR adds a dialog to check for updates and integrates it with the updater service (the widget will be added in the next PR).

This first commit is most important one. It implements updater methods that will be used by the UI. Unfortunately, I found some bugs in `electron-updater`, so I had to add extra logic to resolve them. I've left comments in `AppUpdater.ts` to explain what wasn't working and how those issues were addressed. If anything is unclear, please let me know.

While testing, I also realized that the UI in the multi-cluster scenario may be too complex when we show errors/warning about unreachable clusters (although it seemed fine in Storybook to me). I'll address this separately by reducing alerts and making inline errors more prominent.

### How to test it (on macOS)
I prepared two dev builds with different versions (`19.0.0-dev.connect.updates.1` and `19.0.0-dev.connect.updates.2`) to test the updates.
1. Globally set the base URL env var: `launchctl setenv TELEPORT_CDN_BASE_URL https://cdn.cloud.gravitational.io`.
2. Set your cluster client tools version: `tctl autoupdate client-tools target 19.0.0-dev.connect.updates.2`.
3. Download `19.0.0-dev.connect.updates.1` build from https://cdn.cloud.gravitational.io/Teleport%20Connect-19.0.0-dev.connect.updates.1.dmg.
4. Move the app to Applications and start it -> App menu -> Check for updates.

### CODE REVIEW

File: web/packages/teleterm/src/ui/AppUpdater/AppUpdaterContext.tsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+
+interface AppUpdaterContext {
+  updateEvent: AppUpdateEvent;
+  openDialog(): void;
+}
+
+const AppUpdaterContext = createContext<AppUpdaterContext>(null);
+
+export function AppUpdaterContextProvider(props: PropsWithChildren) {
```
Comment (gzdunek) [inline web/packages/teleterm/src/ui/AppUpdater/AppUpdaterContext.tsx:39]: At first, I wanted to expose all updater methods through this context to keep everything in one place, but well, I'd just re-export functions of `mainProcessClient`.
I wasn't sure if this was a good idea, so in the end, I didn't add them here and the callsites will rely both on this context and `mainProcessClient`.

File: web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx
Comment (ravicious) [inline web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx:1]: I followed your instructions and the process itself was pretty nice. I had to choose a cluster first because of incompatible versions (Cloud cluster on v17 and the other cluster on v19.0.0-dev.connect.updates.2). Maybe the only thing that surprised me was that the download started the moment I selected `cluster.mirrors.link` as the cluster. I expected there to be some kind of a confirmation first, but maybe the way it is implemented right now is okay – there's fewer clicks and the download can always be canceled.

However, after updating the app to .2, I consistently see this error in the modal:

<details>
<summary>main.log</summary>

```
[04-08-25 18:09:58] [AppUpdater] info: Checking for update
[04-08-25 18:09:58] [tshd] info: send teleport.lib.teleterm.auto_update.v1.AutoUpdateService GetClusterVersions {}
[04-08-25 18:09:58] [tshd] info: receive teleport.lib.teleterm.auto_update.v1.AutoUpdateService GetClusterVersions {"reachableClusters":{"0":{"clusterUri":"/clusters/enterprise-local","toolsAutoUpdate":false,"toolsVersion":"19.0.0-dev","minToolsVersion":"18.0.0-aa"},"1":{"clusterUri":"/clusters/cluster.mirrors.link","toolsAutoUpdate":true,"toolsVersion":"19.0.0-dev.connect.updates.2","minToolsVersion":"17.0.0-aa"},"2":{"clusterUri":"/clusters/ravicious.cloud.gravitational.io","toolsAutoUpdate":false,"toolsVersion":"17.5.6","minToolsVersion":"16.0.0-aa"}},"unreachableClusters":{}}
[04-08-25 18:09:58] [AppUpdater] info: Resolved auto updates status {"options":{"clusters":[{"clusterUri":"/clusters/enterprise-local","toolsAutoUpdate":false,"toolsVersion":"19.0.0-dev","minToolsVersion":"18.0.0-aa","otherCompatibleClusters":["/clusters/cluster.mirrors.link"]},{"clusterUri":"/clusters/cluster.mirrors.link","toolsAutoUpdate":true,"toolsVersion":"19.0.0-dev.connect.updates.2","minToolsVersion":"17.0.0-aa","otherCompatibleClusters":["/clusters/enterprise-local"]},{"clusterUri":"/clusters/ravicious.cloud.gravitational.io","toolsAutoUpdate":false,"toolsVersion":"17.5.6","minToolsVersion":"16.0.0-aa","otherCompatibleClusters":["/clusters/cluster.mirrors.link"]}],"unreachableClusters":[],"managingClusterUri":"/clusters/cluster.mirrors.link"},"enabled":true,"version":"19.0.0-dev.connect.updates.2","source":"managing-cluster"}
[04-08-25 18:09:58] [tshd] info: send teleport.lib.teleterm.auto_update.v1.AutoUpdateService GetDownloadBaseUrl {}
[04-08-25 18:09:58] [tshd] info: receive teleport.lib.teleterm.auto_update.v1.AutoUpdateService GetDownloadBaseUrl {"baseUrl":"https://cdn.cloud.gravitational.io"}
[04-08-25 18:09:58] [AppUpdater] info: Update for version 19.0.0-dev.connect.updates.2 is not available (latest version: 19.0.0-dev.connect.updates.2, downgrade is allowed).
[04-08-25 18:09:58] [AppUpdater] error: Error: Error: Please check update first
    at MacUpdater.downloadUpdate (/Applications/Teleport Connect.app/Contents/Resources/app.asar/build/app/main/index.js:22189:24)
    at AppUpdater.download (/Applications/Teleport Connect.app/Contents/Resources/app.asar/build/app/main/index.js:24003:52)
    at AppUpdater.doCheckForUpdates (/Applications/Teleport Connect.app/Contents/Resources/app.asar/build/app/main/index.js:23990:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async AppUpdater.checkForUpdates (/Applications/Teleport Connect.app/Contents/Resources/app.asar/build/app/main/index.js:23964:7)
    at async Session.<anonymous> (node:electron/js2c/browser_init:2:107024)
[04-08-25 18:09:58] [AppUpdater] error: Failed to download update. Error: Please check update first
    at MacUpdater.downloadUpdate (/Applications/Teleport Connect.app/Contents/Resources/app.asar/build/app/main/index.js:22189:24)
    at AppUpdater.download (/Applications/Teleport Connect.app/Contents/Resources/app.asar/build/app/main/index.js:24003:52)
    at AppUpdater.doCheckForUpdates (/Applications/Teleport Connect.app/Contents/Resources/app.asar/build/app/main/index.js:23990:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async AppUpdater.checkForUpdates (/Applications/Teleport Connect.app/Contents/Resources/app.asar/build/app/main/index.js:23964:7)
    at async Session.<anonymous> (node:electron/js2c/browser_init:2:107024)
```

</details>

error-app-updates

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts [hunk 2]
```diff
@@ -89,15 +107,202 @@ export class AppUpdater {
...
+        // update directory for Connect.
+        if (autoUpdater instanceof MacUpdater && app.isPackaged) {
+          const squirrelMacPath = path.join(
+            os.homedir(),
+            'Library',
+            'Caches',
+            'gravitational.teleport.connect.ShipIt'
+          );
+          try {
+            rmSync(squirrelMacPath, {
```
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts]: At first I wanted to suggest calling `event.preventDefault()` and then using an async function so that we can get rid of `rmSync`.

This alone wouldn't be that helpful though, because we already have a handler with `will-quit` that also calls the same method on the event.

Would it be possible to instead perform this removal from within the existing `will-quit` handler? This way we can orchestrate everything in one place and not worry about blocking the main thread or two handlers racing each other.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts [hunk 3]
```diff
@@ -89,15 +107,202 @@ export class AppUpdater {
...
+              recursive: true,
+              force: true,
+              maxRetries: 2,
+            });
+          } catch {
+            /* empty */
```
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts]: Should we at least `console.log` the error? If this logic is moved to the `will-quit` handler in main.ts, then we can even send it to `logger`.

`console.log` would be the last resort since no one will ever think to start Connect from console just to see check for an error from this.

Edit: Oh I see that `logger` can be accessed even from here.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts (hunk 2 above)
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts]: Is it possible that this races with Squirrel? Or in other words, when does exactly Squirrel do its thing? Once the app completely quits? Is code from `will-quit` guaranteed to finish until Squirrel performs an update?

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts [hunk 4]
```diff
@@ -89,15 +107,202 @@ export class AppUpdater {
...
+  /**
+   * Checks for app updates.
+   *
+   * This method enhances the standard autoUpdater.checkForUpdates() by adding
+   * the following behaviors:
+   * - It allows update checks during an ongoing download process.
+   * If a new update is found (or no update is available), the current download
+   * is canceled.
+   * - If an update is checked after a download has completed, the updater
+   * automatically transitions to the `update-downloaded` state.
```
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts]: What is the default behavior for this?

File: web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx
Comment (gzdunek) [reply web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx:1]: >Maybe the only thing that surprised me was that the download started the moment I selected cluster.mirrors.link as the cluster. I expected there to be some kind of a confirmation first, but maybe the way it is implemented right now is okay – there's fewer clicks and the download can always be canceled.

Yeah, it can be a bit surprising, probably because the list doesn't indicate whether selecting an option will trigger a download or if you're already on that version.
I feel like this UI part could be more refined, but on the other hand, only a small number of users are likely to see it anyway.

>However, after updating the app to .2, I consistently see this error in the modal:

Oops, I shouldn't try to downloaded the update if there are no updates. Fixed.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts (hunk 2 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/appUpdater/appUpdater.ts]: Ah right, I completely forgot that we already have a `will-quit` handler. It's even simpler now, as I can move that code into the existing `appUpdater.dispose` which is called on `will-quit`.

Comment (gzdunek) [reply web/packages/teleterm/src/services/appUpdater/appUpdater.ts]: The auto-installation begins on `quit` event.

I realized that trying to remove the entire `ShipIt` directory might not have been the best idea. For some reason, removing `app.asar` in the unzipped app fails:
>ENOTDIR: not a directory, rmdir '/Users/grzegorz/Library/Caches/gravitational.teleport.connect.ShipIt/update.0uIISpb/Teleport Connect.app/Contents/Resources/app.asar'

But instead we can remove just `gravitational.teleport.connect.ShipIt/ShipItState.plist` file that controls the auto-installation.

When it comes to racing with Squirrel, I'll first describe how applying the update works:
1. The update download completes on the Electron Updater side, and a corresponding event is emitted.
2. The downloaded file is then copied from Electron Updater’s `/pending` directory to a path like `~/Library/Caches/gravitational.teleport.connect.ShipIt/update.3uGWT0G/f0f20676162197f34e51545e1880ddd935d314005425ce86f90c132f2a5e5a52e3e6d5b0e83454d8e582ed8e06e1fc512c2386918d5c876b77578d922aa70ae2.zip`.
3. Squirrel immediately begins unzipping the file in the background. However, the UI no longer shows the "downloading" state. This is a bit buggy on its own. If the app is manually closed right after the download completes, the unzip process won’t finish, and the update won't be applied.
4. If the unzipping does finish, a `gravitational.teleport.connect.ShipIt/ShipItState.plist` file is created. In this case, closing the app will apply the update successfully.

Technically, a race condition could occur if the unzipping completes between the `will-quit` and `quit` events, and we check for updates again (receiving no updates) before `will-quit`.
However, this scenario doesn't seem likely to happen? But even if it does, the worst thing that can happen is that we will apply an outdated update.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts (hunk 4 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/appUpdater/appUpdater.ts]: If auto download is enabled in Electron Updater, it will attempt to download the update again, so the updater will transition to the `update-downloaded` state too.
But if it's not, then the updater will stay in the `update-available` state.

I clarified that only that second case has been changed.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts [hunk 5]
```diff
@@ -107,11 +304,48 @@ export class AppUpdater {
...
   }
+
+  /**
+   * Workaround to prevent installing outdated updates.
+   * electron-updater lacks support for this: once an update is downloaded,
+   * it will be installed on quit—even if subsequent update checks report
+   * no new updates.
+   */
+  private async preventInstallingOutdatedUpdates(): Promise<void> {
+    if (!this.isUpdateDownloaded) {
```
Comment (ryanclark) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts]: Thoughts on inverting this and returning early to bring the rest of the function less indented?

**Review by ryanclark — APPROVED** [review]

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts [hunk 6]
```diff
@@ -89,15 +107,202 @@ export class AppUpdater {
...
+      this.logger.info('Download already in progress.');
+      return this.downloadPromise;
+    }
+
+    this.downloadPromise = autoUpdater.downloadUpdate();
+    try {
+      await this.downloadPromise;
+      this.isUpdateDownloaded = true;
+    } catch (error) {
+      this.logger.error('Failed to download update.', error);
```
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts:242]: Does this error end up being surfaced to the user? I assume this might happen through `autoUpdater`. It'd be good to leave a comment about how this error gets surfaced.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts [hunk 7]
```diff
@@ -89,15 +107,202 @@ export class AppUpdater {
...
+      this.isUpdateDownloaded = false;
+      return false;
+    }
+
+    // Due to a bug in electron-updater, we can't cancel downloads using cancellation
+    // token passed to autoUpdater.download().
+    // Repeatedly starting and canceling downloads causes the updater to go
+    // into a broken state where it becomes unresponsive.
+    // To avoid this, we instead close the network connections to abort
+    // the current download.
```
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts:260]: This is so cursed. 😂 It's surprising that `netSession` is even available through `autoUpdater`.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts [hunk 8]
```diff
@@ -89,15 +107,202 @@ export class AppUpdater {
...
+    try {
+      await autoUpdater.netSession.closeAllConnections();
+      await this.downloadPromise;
+      return false;
+    } catch {
+      return true;
```
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts:266]: Is there a way for this `catch` to be more specific rather than returning `true` for all errors? Or maybe at least wrap just `this.downloadPromise` in `try catch`?

I assume that we don't care about the error from `this.downloadPromise` and just want to wait until the promise fails – a very broad `try catch` like that is understandable in this case. But we expect `closeAllConnections` to not throw, right? If it fails, it feels like the whole method should fail too.

File: web/packages/teleterm/src/mainProcess/mainProcessClient.ts [hunk 9]
```diff
@@ -199,5 +202,40 @@ export default function createMainProcessClient(): MainProcessClient {
...
+    downloadAppUpdate() {
+      return ipcRenderer.invoke(MainProcessIpc.DownloadAppUpdate);
+    },
+    cancelAppUpdateDownload() {
+      return ipcRenderer.invoke(MainProcessIpc.CancelAppUpdateDownload);
+    },
+    quitAndInstallAppUpdate() {
+      return ipcRenderer.invoke(MainProcessIpc.QuiteAndInstallAppUpdate);
+    },
+    changeAppUpdatesManagingCluster(clusterUri) {
```
Comment (ravicious) [inline web/packages/teleterm/src/mainProcess/mainProcessClient.ts:219]: Is it a problem if a rogue renderer changes the managing cluster to a cluster that is not in the app?

It doesn't feel like it should be a problem because `getClusterVersions` always works on data returned from tshd. So the worst thing that this IPC can cause is simply forcing the user to pick a managing cluster again.

File: web/packages/teleterm/src/mainProcess/mainProcessClient.ts [hunk 10]
```diff
@@ -199,5 +202,40 @@ export default function createMainProcessClient(): MainProcessClient {
...
+        {
+          clusterUri,
+        }
+      );
+    },
+    subscribeToAppUpdateEvents: listener => {
+      const ipcListener = (_, updateEvent: AppUpdateEvent) => {
+        // Deserialize the error.
+        if (updateEvent.kind === 'error') {
+          updateEvent.error = ensureError(updateEvent.error);
```
Comment (ravicious) [inline web/packages/teleterm/src/mainProcess/mainProcessClient.ts]: If you control both ends (one that serializes the error and one that deserializes it), I imagine in a perfect world it'd be better to have a specialized function for this which handles the exact serialization shape used by the other end. Depending on `ensureError` to do this job feels a bit off but I think it's acceptable. ;f

**Review by ravicious — COMMENTED** [review]
I managed to look through the new commits and the old ones until "Expose new APIs through Electron IPC", I'll try to finish the review by the end of this day.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts (hunk 5 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/appUpdater/appUpdater.ts]: Absolutely, inverted.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts (hunk 6 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/appUpdater/appUpdater.ts:242]: Errors that happen in `autoUpdater.checkForUpdates` or `autoUpdater.download` are emitted as events. I added comments.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts (hunk 7 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/appUpdater/appUpdater.ts:260]: Yeah :/ 
I suspect something is wrong with closing the connection on cancel, so the old connection is not closed and a new one cannot be opened.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts (hunk 8 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/appUpdater/appUpdater.ts:266]: Makes sense, I moved `closeAllConnections` out of try catch.

File: web/packages/teleterm/src/mainProcess/mainProcessClient.ts (hunk 9 above)
Comment (gzdunek) [reply web/packages/teleterm/src/mainProcess/mainProcessClient.ts:219]: Correct, `resolveAutoUpdatesStatus` completely ignores clusters that do not exist. 
So if you have a single cluster and `changeAppUpdatesManagingCluster` sets a different cluster, you won't notice anything. 
If you had a cluster selected, you will need to select it again.

File: web/packages/teleterm/src/mainProcess/mainProcessClient.ts (hunk 10 above)
Comment (gzdunek) [reply web/packages/teleterm/src/mainProcess/mainProcessClient.ts]: I added functions to serialize/deserialize errors. I got rid of serialization in `appUpdater` (which doesn't need to be aware of it), now this happens before sending the event through IPC.

File: web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+  }, [checkForAppUpdates]);
+
+  const platform = useMemo(() => {
+    return appContext.mainProcessClient.getRuntimeSettings().platform;
+  }, [appContext.mainProcessClient]);
+
+  return (
+    <DialogConfirmation
+      open={!props.hidden}
+      keepInDOMAfterClose
```
Comment (ravicious) [inline web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx:54]: We need this prop on pretty much every dialog in Connect, don't we?

I'm not saying we should do it within this PR, but at some point we should add Connect-specific `DialogConfirmation` just so that we don't have to remember about this. Plus this JSDoc should explain when it is useful because I already managed to partially forgot what the use case was and had to spend a few seconds trying to remember the original problem that we solved with it. 😅

https://github.com/gravitational/teleport/blob/511a37de71f197cec5f421836e4c28e545f79491/web/packages/design/src/Modal/Modal.tsx#L33-L37

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts [hunk 12]
```diff
@@ -59,6 +59,7 @@ export class AppUpdater {
   private checkForUpdatesPromise: Promise<void> | undefined;
   private downloadPromise: Promise<string[]> | undefined;
   private isUpdateDownloaded = false;
+  private forceNoAutoDownload = false;
```
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts:65]: This field doesn't seem to be used. I guess it's maybe a mistake and it's supposed to be set at some point when `noAutoDownload` is used? Because I don't see `noAutoDownload` being used either.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts [hunk 13]
```diff
@@ -284,6 +288,26 @@ export class AppUpdater {
...
+    }
+
+    // checkForUpdates will discard any update triggered by the removed managing
+    // cluster. If a different update is found, do not download it automatically.
+    // Currently, updates aren't checked in the background, and on Windows and Linux,
+    // users may be surprised by an admin prompt when there is an update to install
+    // after quitting the app.
+    // We may revisit this behavior in the future. For example, if we introduce
+    // a UI notification indicating that there's an update to be installed.
+    await this.checkForUpdates({ noAutoDownload: true });
```
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts:303]: Is there any way we can create tests for this? It feels quite brittle, as in we don't directly discard any updates here but rather we depend on `this.checkForUpdates` doing this somewhere underneath when called with `noAutoDownload`. It'd be at least nice to cover that it in fact does this when `noAutoDownload` is used.

Is it possible to mock/fake `autoUpdater`? The logic in this file is already rather complex and we just don't have any tests for it, meaning that even this edge case here would need to be covered by the test plan.

File: web/packages/teleterm/src/ui/ClusterLogout/useClusterLogout.ts [hunk 14]
```diff
@@ -30,6 +30,9 @@ export function useClusterLogout({
   const ctx = useAppContext();
   const [{ status, statusText }, removeCluster] = useAsync(async () => {
     await ctx.clustersService.logout(clusterUri);
+    await ctx.mainProcessClient.maybeRemoveAppUpdatesManagingCluster(
+      clusterUri
+    );
```
Comment (ravicious) [inline web/packages/teleterm/src/ui/ClusterLogout/useClusterLogout.ts]: Does this line in combination with the following one mean that when the user logs out of a cluster, the whole logout procedure won't be finished until the app finishes a check for updates?

https://github.com/gravitational/teleport/blob/511a37de71f197cec5f421836e4c28e545f79491/web/packages/teleterm/src/services/appUpdater/appUpdater.ts#L270-L277

If that's the case, I feel like logging out of a cluster shouldn't be allowed to wait for an update check of shouldn't be forced to fail if something goes wrong with auto-updates. Perhaps it should be done in a background instead and possible errors could be logged or shown as a notification? I assume the user won't care about the error until they actually open the update modal.

File: web/packages/teleterm/src/services/appUpdater/clientToolsUpdateProvider.ts [hunk 15]
```diff
@@ -68,7 +68,7 @@ export class ClientToolsUpdateProvider extends Provider<UpdateInfo> {
     }
 
     const { baseUrl, version } = clientTools;
-    const fileUrl = `https://${baseUrl}/${makeDownloadFilename(this.nativeUpdater, version)}`;
+    const fileUrl = `${baseUrl}/${makeDownloadFilename(this.nativeUpdater, version)}`;
```
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/clientToolsUpdateProvider.ts:71]: This is read from the env var or otherwise uses our default value for non-OSS versions, right?

I tried providing a wrong URL through the env var (without https) and I got the following error. I don't think it should block the release of this feature, but perhaps we should note it down as tech debt somewhere that the Go implementation of AutoUpdateService should validate the scheme of the URL so that hopefully we can return a more meaningful error.

```
MAIN [tshd] info: send teleport.lib.teleterm.auto_update.v1.AutoUpdateService GetDownloadBaseUrl {}
MAIN [tshd] info: receive teleport.lib.teleterm.auto_update.v1.AutoUpdateService GetDownloadBaseUrl {"baseUrl":"cdn.cloud.gravitational.io"}
MAIN [AppUpdater] error: Error: TypeError: Failed to parse URL from cdn.cloud.gravitational.io/Teleport Connect-19.0.0-dev.connect.updates.2-mac.zip.sha512
    at node:internal/deps/undici/undici:13510:13
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async fetchChecksum (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:23864:20)
    at async updateProvider.getLatestVersion (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:23819:20)
    at async MacUpdater.getUpdateInfoAndProvider (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:22143:15)
    at async MacUpdater.doCheckForUpdates (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:22156:22)
    at async AppUpdater.doCheckForUpdates (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:23956:20)
    at async AppUpdater.checkForUpdates (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:23947:7)
    at async Session.<anonymous> (node:electron/js2c/browser_init:2:107024)
MAIN [AppUpdater] error: Failed to check for updates. TypeError: Failed to parse URL from cdn.cloud.gravitational.io/Teleport Connect-19.0.0-dev.connect.updates.2-mac.zip.sha512
    at node:internal/deps/undici/undici:13510:13
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async fetchChecksum (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:23864:20)
    at async updateProvider.getLatestVersion (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:23819:20)
    at async MacUpdater.getUpdateInfoAndProvider (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:22143:15)
    at async MacUpdater.doCheckForUpdates (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:22156:22)
    at async AppUpdater.doCheckForUpdates (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:23956:20)
    at async AppUpdater.checkForUpdates (/Users/rav/src/teleport/web/packages/teleterm/build/app/main/index.js:23947:7)
    at async Session.<anonymous> (node:electron/js2c/browser_init:2:107024)
```

invalid-url

**Review by ravicious — COMMENTED** [review]
I'll do some light QA tomorrow as so far I've mostly just read the code.

File: web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx
Comment (ravicious) [inline web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx:1]: Improvement for the future: It'd be nice if after downloading the update there was a button to abort the update.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts
Comment (ravicious) [inline web/packages/teleterm/src/services/appUpdater/appUpdater.ts:1]: The updater seems to work well on all platforms. It's going to be a pain to set the env var on Linux though, we'll need to add some steps on how to find the .desktop file. https://github.com/gravitational/teleport/pull/56493#discussion_r2200236436

File: web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx (hunk 11 above)
Comment (gzdunek) [reply web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx:54]: Yeah, we need this in every dialog, I agree that we should refactor it 🫠

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts (hunk 12 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/appUpdater/appUpdater.ts:65]: Oops, I likely overlooked it while preparing commits. 
Its value should be set on each check of `doCheckForUpdates`. I added tests for it.

File: web/packages/teleterm/src/services/appUpdater/appUpdater.ts (hunk 13 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/appUpdater/appUpdater.ts:303]: I thought it would a pain to mock `autoUpdater`, but it wasn't that bad! 
I added tests for updates auto download and discarding them when the latest check returns a new update or no update, as this part feels the most brittle.

File: web/packages/teleterm/src/ui/ClusterLogout/useClusterLogout.ts (hunk 14 above)
Comment (gzdunek) [reply web/packages/teleterm/src/ui/ClusterLogout/useClusterLogout.ts]: Great catch, we shouldn't wait for it.

File: web/packages/teleterm/src/services/appUpdater/clientToolsUpdateProvider.ts (hunk 15 above)
Comment (gzdunek) [reply web/packages/teleterm/src/services/appUpdater/clientToolsUpdateProvider.ts:71]: >This is read from the env var or otherwise uses our default value for non-OSS versions, right?

Yes.
I left a TODO item for adding validation for it.

File: web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx
Comment (gzdunek) [reply web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx:1]: I was considering it, but I was afraid of making auto updates too complex. Also, I don't think I know any app/system that would allow it 😅 
Do you see a specific scenario where this would be useful?

Comment (ravicious) [reply web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx:1]: My original comment might have been a result of a developer brain. I was testing the feature and it downloaded an update that I didn't really need, but I couldn't skip it once it's been downloaded.

But this got me thinking: the newest version of my editor is currently broken. The updater fortunately offers a "Skip this update" feature (I think it's supported OOTB in one of the popular macOS updaters). What if we release a bugged version of Connect? The cluster admin would probably need to turn off auto-updates, but why would they need to do that if there's a problem only with Connect and not tsh? Or say the bug affects only a specific platform. As the end user, can I skip it somehow? Is setting an env var the only available workaround?

**Review by ravicious — APPROVED** [review]

File: web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx
Comment (gzdunek) [reply web/packages/teleterm/src/ui/ModalsHost/modals/AppUpdates.tsx:1]: For the CLI tools, the official workaround is the env var:
>An environment variable TELEPORT_TOOLS_VERSION can be used as an emergency workaround for a known issue, pinning to a specific version in CI/CD, for debugging, or for manual updates.

The same one will be available in Connect. I know it's a bit inconvenient for desktop apps, but I also feel it's acceptable for the updater v1.
I can open an issue to add a UI option in the future that will allow users to skip an update or maybe even revert to a previous version.
